compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java

```diff
@@ -1470,34 +1476,64 @@ void generateMapLoadIns(BIRNonTerminator.FieldAccess mapLoadIns) {
 
         // visit key_expr
         this.loadVar(mapLoadIns.keyOp.variableDcl);
-
+        BType targetType = mapLoadIns.lhsOp.variableDcl.type;
+        this.mv.visitTypeInsn(CHECKCAST, B_STRING_VALUE);
+        boolean shouldUnbox = true;
         if (varRefType.tag == TypeTags.JSON) {
-
             if (mapLoadIns.optionalFieldAccess) {
-                this.mv.visitTypeInsn(CHECKCAST, B_STRING_VALUE);
-                this.mv.visitMethodInsn(INVOKESTATIC, JSON_UTILS, "getElementOrNil",
-                                        JSON_GET_ELEMENT,
-                                        false);
+                this.mv.visitMethodInsn(INVOKESTATIC, JSON_UTILS, "getElementOrNil", JSON_GET_ELEMENT, false);
             } else {
-                this.mv.visitTypeInsn(CHECKCAST , B_STRING_VALUE);
                 this.mv.visitMethodInsn(INVOKESTATIC, JSON_UTILS, "getElement", JSON_GET_ELEMENT, false);
             }
         } else {
             if (mapLoadIns.fillingRead) {
                 this.mv.visitMethodInsn(INVOKEINTERFACE, MAP_VALUE, "fillAndGet",
                         PASS_OBJECT_RETURN_OBJECT, true);
             } else {
-                this.mv.visitMethodInsn(INVOKEINTERFACE, MAP_VALUE, "get",
-                        PASS_OBJECT_RETURN_OBJECT, true);
+                shouldUnbox = generateMapGet(varRefType, targetType);
             }
         }
 
         // store in the target reg
-        BType targetType = mapLoadIns.lhsOp.variableDcl.type;
-        jvmCastGen.addUnboxInsn(this.mv, targetType);
+        if (shouldUnbox) {
+            jvmCastGen.addUnboxInsn(this.mv, targetType);
+        }
         this.storeToVar(mapLoadIns.lhsOp.variableDcl);
     }
 
+    boolean generateMapGet(BType mapType, BType expectedType) {
+        if (mapType.getKind() != TypeKind.RECORD) {
+            this.mv.visitMethodInsn(INVOKEINTERFACE, MAP_VALUE, "get", PASS_OBJECT_RETURN_OBJECT, true);
+            return true;
+        }
+        return switch (expectedType.getKind()) {
+            case INT -> {
+                this.mv.visitMethodInsn(INVOKEVIRTUAL, MAP_VALUE_IMPL, "getUnboxedIntValue",
+                        PASS_B_STRING_RETURN_UNBOXED_LONG, false);
+                yield false;
+            }
+            case FLOAT -> {
+                this.mv.visitMethodInsn(INVOKEVIRTUAL, MAP_VALUE_IMPL, "getUnboxedFloatValue",
+                        PASS_B_STRING_RETURN_UNBOXED_DOUBLE, false);
+                yield false;
+            }
+            case STRING -> {
+                this.mv.visitMethodInsn(INVOKEVIRTUAL, MAP_VALUE_IMPL, "getStringValue", PASS_B_STRING_RETURN_B_STRING,
+                        false);
+                yield true;
+            }
+            case BOOLEAN -> {
+                this.mv.visitMethodInsn(INVOKEVIRTUAL, MAP_VALUE_IMPL, "getUnboxedBooleanValue",
```

Comment:
What is the reason to introduce a new function `getUnboxedBooleanValue`? Can't we utilize the `getBooleanValue` function in the `MapValueImp` class instead?

Reply:
`getBooleanValue` boxes `boolean` values to `Boolean`. Since in ballerina represent primitive values like `boolean` using java primitives we then have to unbox it using by casting to `Boolean` which jvm optimize away as a no-op cast and then calling [`booleanValue()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Boolean.html#booleanValue()). `getUnboxedBooleanValue`, directly return the `boolean` value (since we store record fields as primitives) letting us avoid this unboxing step.